packages/appsync-modelgen-plugin/src/visitors/appsync-typescript-visitor.ts

```diff
@@ -159,29 +157,36 @@ export class AppSyncModelTypeScriptVisitor<
    */
   protected generateModelDeclaration(modelObj: CodeGenModel, isDeclaration: boolean = true, isModelType: boolean = true): string {
     const modelName = this.generateModelTypeDeclarationName(modelObj);
-    const modelDeclarations = new TypeScriptDeclarationBlock()
-      .asKind('class')
-      .withFlag({ isDeclaration })
-      .withName(modelName)
-      .export(true);
-
-      let readOnlyFieldNames: string[] = [];
-      let modelMetaDataFormatted: string | undefined;
-      let modelMetaDataDeclaration: string = '';
+    const eagerModelDeclaration = new TypeScriptDeclarationBlock().asKind('type').withName(`Eager${modelName}`);
+    const lazyModelDeclaration = new TypeScriptDeclarationBlock().asKind('type').withName(`Lazy${modelName}`);
+
+    let readOnlyFieldNames: string[] = [];
+    let modelMetaDataFormatted: string | undefined;
+    let modelMetaDataDeclaration: string = '';
     //Add new model meta field when custom primary key is enabled
     if (isModelType && this.isCustomPKEnabled()) {
       //Add new model meta import
       this.BASE_DATASTORE_IMPORT.add(this.MODEL_META_FIELD_NAME);
-      modelDeclarations.addProperty(`[${this.MODEL_META_FIELD_NAME}]`, this.generateModelMetaDataType(modelObj), undefined, 'DEFAULT', {
+      const modelMetaDataType = this.generateModelMetaDataType(modelObj);
+      eagerModelDeclaration.addProperty(`[${this.MODEL_META_FIELD_NAME}]`, modelMetaDataType, undefined, 'DEFAULT', {
         readonly: true,
-        optional: false
+        optional: false,
+      });
+      lazyModelDeclaration.addProperty(`[${this.MODEL_META_FIELD_NAME}]`, modelMetaDataType, undefined, 'DEFAULT', {
+        readonly: true,
+        optional: false,
       });
     }
     modelObj.fields.forEach((field: CodeGenField) => {
-      modelDeclarations.addProperty(this.getFieldName(field), this.getNativeType(field), undefined, 'DEFAULT', {
+      const fieldName = this.getFieldName(field);
+      eagerModelDeclaration.addProperty(fieldName, this.getNativeType(field), undefined, 'DEFAULT', {
         readonly: true,
         optional: field.isList ? field.isListNullable : field.isNullable,
       });
+      lazyModelDeclaration.addProperty(fieldName, this.getNativeType(field, { lazy: true }), undefined, 'DEFAULT', {
+        readonly: true,
+        optional: !this.isModelType(field) && (field.isList ? field.isListNullable : field.isNullable),
```

Comment:
The nullability for lazy/eager is different for model types, in which in lazy mode the model type is always optional. Is this an intentional change?

Reply:
It should be the model type is always _not_ optional. A model with a connection will always have a `AsyncCollection` or `AsyncItem` in the connection field. The `AsyncItem` may resolve to `undefined` and the `AsyncCollection` may resolve to an empty list to indicate the field is empty.

We are using some type mappings to allow the fields to not be supplied when creating an instance.

```
type Parent @model {
  child: Child
}
type Child @model {
  ...
}
```

```js
new Parent({});

// equivalent to

new Parent({ child: new AsyncItem((resolve) => { resolve(undefined); });
```

Reply:
Ah I forget `optional: false` means always non-nullable. Make sense to me